File: app/src/main/java/xyz/sleekstats/completist/view/MyListsFragment.java
package xyz.sleekstats.completist.view;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RadioGroup;

import java.util.List;

import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.subjects.PublishSubject;
import xyz.sleekstats.completist.R;
import xyz.sleekstats.completist.databinding.MovieKeys;
import xyz.sleekstats.completist.model.Genre;
import xyz.sleekstats.completist.model.GenreList;
import xyz.sleekstats.completist.model.MyList;
import xyz.sleekstats.completist.viewmodel.MovieViewModel;

public class MyListsFragment extends Fragment
        implements MyListsAdapter.ItemClickListener, GenresDialog.GenreSelector {

    private RecyclerView mPopularListRV;
    private RecyclerView mSavedListRV;
    private MyListsAdapter mPopularListAdapter;
    private MyListsAdapter mSavedListAdapter;
    private final CompositeDisposable myListsCompositeDisposable = new CompositeDisposable();
    private MovieViewModel movieViewModel;
    private RadioGroup radioGroup1;
    private RadioGroup radioGroup2;
    private PublishSubject<List<MyList>> popularActorsSubject;

    public MyListsFragment() {
        // Required empty public constructor
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_my_lists, container, false);

        radioGroup1 = view.findViewById(R.id.radio_group1);
        radioGroup2 = view.findViewById(R.id.radio_group2);
        view.findViewById(R.id.watched_movies_btn).setOnClickListener(btn -> onListClick(MovieKeys.LIST_WATCHED));
        view.findViewById(R.id.popular_movies_btn).setOnClickListener(btn -> onListClick(MovieKeys.LIST_POPULAR));
        view.findViewById(R.id.genres_btn).setOnClickListener(btn -> onGenreClick());
        view.findViewById(R.id.nowshowing_movies_btn).setOnClickListener(btn -> onListClick(MovieKeys.LIST_NOWPLAYING));
        view.findViewById(R.id.top_movies_btn).setOnClickListener(btn -> onListClick(MovieKeys.LIST_TOPRATED));
        view.findViewById(R.id.scheduled_btn).setOnClickListener(btn -> onListClick(MovieKeys.LIST_QUEUED));

        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (movieViewModel == null) {
            movieViewModel = ViewModelProviders.of(requireActivity()).get(MovieViewModel.class);
        }
        popularActorsSubject = movieViewModel.getPopularActorsSubject();
        myListsCompositeDisposable.add(
                popularActorsSubject.subscribe(this::loadPopularRV,
                        e -> {
                            movieViewModel.finishScrollLoading();
                            Log.e("rxprob", "getPopularActorsSubject loadPopularRV" + e.getMessage());
                        }
                )
        );
        movieViewModel.getPopularActors();
        myListsCompositeDisposable.add(
                movieViewModel.getSavedLists()
                        .subscribe(this::loadSavedRV,
                                e -> Log.e("rxprob", "getSavedLists loadSavedRV" + e.getMessage())
                        )
        );
    }

    private void loadRV(List<MyList> myLists, RecyclerView rv, MyListsAdapter myListsAdapter) {

        if (myListsAdapter == null) {
            myListsAdapter = new MyListsAdapter(myLists);
            myListsAdapter.setClickListener(this);
            rv.setAdapter(myListsAdapter);
        } else {
            myListsAdapter.notifyDataSetChanged();
        }
    }

    private void loadPopularRV(List<MyList> myLists) {

        if (mPopularListRV == null) {
            View rootView = getView();
            if (rootView == null) {
                movieViewModel.finishScrollLoading();
                return;
            }
            mPopularListRV = rootView.findViewById(R.id.pop_lists_rv);
            mPopularListRV.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.HORIZONTAL, false));
            mPopularListRV.addOnScrollListener(new RecyclerView.OnScrollListener() {
                @Override
                public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                    super.onScrolled(recyclerView, dx, dy);
                    if (!recyclerView.canScrollHorizontally(1)) {
                        movieViewModel.queryPopularActors();
                    }
                }
            });
        }

        if (mPopularListAdapter == null) {
            mPopularListAdapter = new MyListsAdapter(myLists);
            mPopularListAdapter.setClickListener(this);
            mPopularListRV.setAdapter(mPopularListAdapter);
        } else {
            mPopularListAdapter.notifyDataSetChanged();
        }
        movieViewModel.finishScrollLoading();
    }

    private void loadSavedRV(List<MyList> myLists) {
        if (mSavedListRV == null) {
            View rootView = getView();
            if (rootView == null) {
                return;
            }
            mSavedListRV = rootView.findViewById(R.id.saved_lists_rv);
            mSavedListRV.setLayoutManager(new LinearLayoutManager(getActivity(), LinearLayoutManager.HORIZONTAL, false));
        }
        loadRV(myLists, mSavedListRV, mSavedListAdapter);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        myListsCompositeDisposable.clear();
    }

    private void onGenreClick() {
        clearRadioGroups();
        myListsCompositeDisposable.add(movieViewModel.getGenreList(true)
                .subscribe(this::openGenresDialog)
        );
    }

    private void openGenresDialog(GenreList genres) {
        if (genres == null) {
            return;
        }
        FragmentManager fragmentManager = (requireActivity()).getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        GenresDialog newFragment = GenresDialog.newInstance(genres);
        newFragment.show(fragmentTransaction, "");
        newFragment.setGenreSelector(this);
    }

    @Override
    public void onListClick(String listID) {
        clearRadioGroups();
        movieViewModel.updateFilms(listID);
    }

    private void clearRadioGroups() {
        if (radioGroup1 != null) {
            radioGroup1.clearCheck();
        }
        if (radioGroup2 != null) {
            radioGroup2.clearCheck();
        }
    }

    @Override
    public void onGenreSelected(Genre genre) {
        movieViewModel.getFilmsByGenre(genre, true);
    }
}

File: app/src/main/java/xyz/sleekstats/completist/view/MyListsFragment.kt
package xyz.sleekstats.completist.view

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RadioGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.PublishSubject
import xyz.sleekstats.completist.R
import xyz.sleekstats.completist.databinding.MovieKeys
import xyz.sleekstats.completist.model.Genre
import xyz.sleekstats.completist.model.GenreList
import xyz.sleekstats.completist.model.MyList
import xyz.sleekstats.completist.view.GenresDialog.GenreSelector
import xyz.sleekstats.completist.viewmodel.MovieViewModel

class MyListsFragment : Fragment(), MyListsAdapter.ItemClickListener, GenreSelector {
    private var mPopularListRV: RecyclerView? = null
    private var mSavedListRV: RecyclerView? = null
    private var mPopularListAdapter: MyListsAdapter? = null
    private val mSavedListAdapter: MyListsAdapter? = null
    private val myListsCompositeDisposable = CompositeDisposable()
    private var radioGroup1: RadioGroup? = null
    private var radioGroup2: RadioGroup? = null
    private var popularActorsSubject: PublishSubject<List<MyList>>? = null

    private val movieViewModel by activityViewModels<MovieViewModel>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_my_lists, container, false)
        view.apply {
            radioGroup1 = findViewById(R.id.radio_group1)
            radioGroup2 = findViewById(R.id.radio_group2)
            findViewById<View>(R.id.watched_movies_btn).setOnClickListener { onListClick(MovieKeys.LIST_WATCHED) }
            findViewById<View>(R.id.popular_movies_btn).setOnClickListener { onListClick(MovieKeys.LIST_POPULAR) }
            findViewById<View>(R.id.genres_btn).setOnClickListener { onGenreClick() }
            findViewById<View>(R.id.nowshowing_movies_btn).setOnClickListener { onListClick(MovieKeys.LIST_NOWPLAYING) }
            findViewById<View>(R.id.top_movies_btn).setOnClickListener { onListClick(MovieKeys.LIST_TOPRATED) }
            findViewById<View>(R.id.scheduled_btn).setOnClickListener { onListClick(MovieKeys.LIST_QUEUED) }
        }
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        popularActorsSubject = movieViewModel.popularActorsSubject
        popularActorsSubject?.let {
            myListsCompositeDisposable.add(
                    it.subscribe({ myLists: List<MyList> -> loadPopularRV(myLists) }
                    ) { e: Throwable ->
                        movieViewModel.finishScrollLoading()
                        Log.e("rxprob", "getPopularActorsSubject loadPopularRV" + e.message)
                    }
            )
        }
        movieViewModel.popularActors
        myListsCompositeDisposable.add(
                movieViewModel.savedLists.subscribe({ myLists: List<MyList> -> loadSavedRV(myLists) }
                ) { e: Throwable -> Log.e("rxprob", "getSavedLists loadSavedRV" + e.message) }
        )

    }

    private fun loadRV(myLists: List<MyList>, rv: RecyclerView?, adapter: MyListsAdapter?) {
        var myListsAdapter = adapter
        if (myListsAdapter == null) {
            myListsAdapter = MyListsAdapter(myLists)
            myListsAdapter.setClickListener(this)
            rv?.adapter = myListsAdapter
        } else {
            myListsAdapter.notifyDataSetChanged()
        }
    }

    private fun loadPopularRV(myLists: List<MyList>) {
        if (mPopularListRV == null) {
            val rootView = view
            if (rootView == null) {
                movieViewModel.finishScrollLoading()
                return
            }
            mPopularListRV = rootView.findViewById(R.id.pop_lists_rv)
            mPopularListRV?.layoutManager = LinearLayoutManager(activity, LinearLayoutManager.HORIZONTAL, false)
            mPopularListRV?.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    if (!recyclerView.canScrollHorizontally(1)) {
                        movieViewModel.queryPopularActors()
                    }
                }
            })
        }
        if (mPopularListAdapter == null) {
            mPopularListAdapter = MyListsAdapter(myLists)
            mPopularListAdapter?.setClickListener(this)
            mPopularListRV?.adapter = mPopularListAdapter
        } else {
            mPopularListAdapter?.notifyDataSetChanged()
        }
        movieViewModel.finishScrollLoading()
    }

    private fun loadSavedRV(myLists: List<MyList>) {
        if (mSavedListRV == null) {
            val rootView = view ?: return
            mSavedListRV = rootView.findViewById(R.id.saved_lists_rv)
            mSavedListRV?.layoutManager = LinearLayoutManager(activity, LinearLayoutManager.HORIZONTAL, false)
        }
        loadRV(myLists, mSavedListRV, mSavedListAdapter)
    }

    override fun onDetach() {
        super.onDetach()
        myListsCompositeDisposable.clear()
    }

    private fun onGenreClick() {
        clearRadioGroups()
        myListsCompositeDisposable.add(
                movieViewModel.getGenreList(true).subscribe({ genres: GenreList? -> openGenresDialog(genres) }
                ) { e: Throwable -> Log.e("rxprob", "getSavedLists loadSavedRV" + e.message) }
        )
    }

    private fun openGenresDialog(genres: GenreList?) {
        if (genres == null) {
            return
        }
        val fragmentManager = requireActivity().supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()
        val newFragment = GenresDialog.newInstance(genres)
        newFragment.show(fragmentTransaction, "")
        newFragment.setGenreSelector(this)
    }

    override fun onListClick(movieID: String?) {
        clearRadioGroups()
        movieViewModel.updateFilms(movieID)
    }

    private fun clearRadioGroups() {
        radioGroup1?.clearCheck()
        radioGroup2?.clearCheck()
    }

    override fun onGenreSelected(genre: Genre?) {
        genre?.let {
            movieViewModel.getFilmsByGenre(genre, true)
        }
    }
}